### PR TITLE
Rag demo sql for using RAG and AIGC to write a report about why customer don't like pizza and how to improve

### PR DESCRIPTION
**Pull request by Wang-Yong2018**
Based on #37  request,  I make a branch called rag_demo. 

# The purpose
:  to write a sql file to from end to end demo how to use pgai to generate a food pizza customer don't like reason report which used both chat_completion and embedding feature.

# The  work
changed content:
1. examples/rag_report_demo_why_some_customers_donot_like_pizza.sql
2. docker-compose.yml

# how to use 
## The requirement for  demo:
1. have access to openai models 'gpt-4o-mini','text-embedding-3-small'.  The api_router should not be used as embedding feature do not support. such as operrouter.
2. pgai docker runing normally
## step for demo
1. open the script file examples/rag_report_demo_why_some_customers_donot_like_pizza.sql
2. replace the ai.openai_api_key with your openai_api_key 
3. test if it work in the test part. or directly run all the script
4. run remained or all script
5. show the sql query result 

# The results.
If everything is OK. the demo will show a table with one records as below
image

In Dbeaver, you can see the final report directly at cell value view as below:
image

# Other features
. docker-compose.yml file add at repository root folder  for easy launch pgai docker container in compose mode. 
I used docker-compose for automaticall launch docker and configuration.  The pgai readme docker part could be replace with docker-compose up -d. That's all

Best Regards,
WangYong

### CODE REVIEW

File: docker-compose.yml
Comment (jgpruitt) [inline docker-compose.yml:1]: This file does not seem necessary. Let's remove it.

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 1]
```diff
@@ -0,0 +1,187 @@
+-- Step 1. Preconfig & Check environment
+-- clean the history test table
+drop table if exists  public.t_demo_text_v1 cascade;   -- customer feedback stored here
+drop table if exists public.t_embeddings_v1 cascade;  -- customer feedback embedded 
+drop table if exists  public.t_ai_report cascade;      -- the ai generated business report
+
+-- make sure pgai docker is running normally
+CREATE EXTENSION IF NOT EXISTS ai CASCADE;
+create extension if not exists vector ;
```
Comment (jgpruitt) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:9]: Line 8 will create the vector extension. We don't need to create it separately.

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+-- the data of embedding result
+CREATE TABLE public.t_embeddings_v1 (
+id bigserial NOT NULL,
+text_id text NOT NULL, 
+text_content text NOT NULL, -- it is same as t_demo_text_v1
+model_name text NOT NULL,
+ntoken int4 NULL,
+nlength int4 NULL,
+embedding public.vector NOT NULL,
```
Comment (jgpruitt) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:63]:
```suggestion
embedding public.vector(1536) NOT NULL,
```
specify the dimensions

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		(text_id, text_content, model_name, embedding )
+	select 
+		id, customer_message, model_name, embedding
+	from 
+		tmp;
+
+--- create index for query speed optimization. It is optional for small data demo 
+-- CREATE INDEX ON t_embeddings_v1 USING ivfflat (embedding vector_cosine_ops) WITH (lists='5');
+-- for unknow reason, I can not build index this time. It report Error: column does not have dimensions. 
+-- as small data, skip the index part at this moment
```
Comment (jgpruitt) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:91]: https://platform.openai.com/docs/models/embeddings

`text-embedding-3-small` uses 1,536 dimensions use `vector(1536)` for the datatype in the column definition.

**Review by jgpruitt — CHANGES_REQUESTED** [review]
This is very cool. Thank you for contributing. There are a few items to fix. Could you also clean up the formatting a bit, please?

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 2 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: Thanks for fix it. I also enabled the create index part which disabled early this no dimension issue.

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 3 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: good reference.

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+
+
+
+---- step 2 create customer feedback data table and insert the demo data.
+--- the original customer feedback message to food 
+CREATE TABLE public.t_demo_text_v1 (
+id bigserial NOT NULL,
+title text NOT NULL,
+url text NULL,
```
Comment (jonatas) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:57]: seems `url` is not used, I'd vote for removing to make it easier to read.

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 5]
```diff
@@ -0,0 +1,205 @@
+-- The demo of AI RAG by generate a why customer don't like pizza report.
+--
+-- The input is 3 records of customer comments to pizza food
+-- the output is ai result table with 3 columns:send_message, chat_completion,  final_report
+--    the final report columns is markdown formated business analysis report to why customer don't like pizza
```
Comment (jonatas) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:5]:
```suggestion
--  the final report columns is markdown formatted business analysis report to why customer don't like pizza.
```

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+-- the process/tools:
+	--- vector database: pgai container for run an store the input, output data
```
Comment (jonatas) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:8]:
```suggestion
    --- vector database: pgai container for run an store the input, output data
```

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+--		message, 
+--		openai_chat_complete(
+--			'gpt-4o-mini',
+--			jsonb_build_array(
+--				jsonb_build_object('role', 'user', 'content', message)
+--			)) as chat, 
+--		openai_embed('text-embedding-3-small',message) as embed
+--	from tmp)
+--select * from test_result;
+
```
Comment (jonatas) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:32]: I think these steps could be removed as they're also in the main README 🤔 

```suggestion
```

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql [hunk 8]
```diff
@@ -0,0 +1,205 @@
+-- The demo of AI RAG by generate a why customer don't like pizza report.
+--
+-- The input is 3 records of customer comments to pizza food
+-- the output is ai result table with 3 columns:send_message, chat_completion,  final_report
+--    the final report columns is markdown formated business analysis report to why customer don't like pizza
+
+-- the process/tools:
+	--- vector database: pgai container for run an store the input, output data
+    --- ai model service: openai 'gpt-4o-mini','text-embedding-3-small'
+    --- the rag concept of process
```
Comment (jonatas) [inline examples/rag_report_demo_why_some_customers_donot_like_pizza.sql:10]:
```suggestion
    --- the RAG concept of the process.
```

**Review by jonatas — CHANGES_REQUESTED** [review]
Great work @Wang-Yong2018! Thanks for doing this!

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 2 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: done

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 5 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: done

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 7 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: Sure. Done

File: examples/rag_report_demo_why_some_customers_donot_like_pizza.sql (hunk 8 above)
Comment (Wang-Yong2018) [reply examples/rag_report_demo_why_some_customers_donot_like_pizza.sql]: done

**Comment by Wang-Yong2018** [comment]
add blog under docs folder named pgai_rag_demo_why_customer_donot_like_pizza.md
4 png attachment added as well.

**Review by jgpruitt — CHANGES_REQUESTED** [review]
Your PR has deleted the `ai--*.sql` files in the `sql` directory. We can't do that. Please revert those deletions.

**Comment by Wang-Yong2018** [comment]
> Your PR has deleted the `ai--*.sql` files in the `sql` directory. We can't do that. Please revert those deletions.
@jonatas I am so sorry for that mistake. It was happened by rag_demo branch merge. For some how, the 3 sql file was deleted by mistake. Now it has been reverted.

**Review by jonatas — CHANGES_REQUESTED** [review]
I love this! Great work @Wang-Yong2018! 

I think @contentana will love to port it to our official blog as well 🚀 

A few ideas about naming:

1. Clean up the suffixes and prefixes like `t_` and  `_v1` as they don't add real value to the case.
2. table names could be more meaningful, like `T_DEMO_TEXT_V1` can be `pizza_reviews` or simply `reviews` and the `subject` can be `Pizza`. 
3. the `T_AI_REPORT` could also be like `ai_reports`

```sql
INSERT INTO reviews (product, message) VALUES
	 ('Pizza','The best pizza I''ve ever eaten. The sauce was so tangy!'),
```

Not asking you to change but just thinking loud that I love the potential of this example, and I'm already seeing we built something like an `ai_instructions` that could be a building block to not spread instructions around the entire database or keep it inside the functions. Something that could be:

```sql
insert into daily_reports (day, subject, content) as
with top as (
   select time_bucket('1 day', time),
    string_agg(content) as content
    FROM reviews where subject = 'Pizza Reviews' 
    ORDER BY embedding <-> eqt.embedding as similarity
    LIMIT 3
  )
 select build_report('daily operational issues report', top.content);
 ```

File: docs/pgai_rag_demo_why_customer_donot_like_pizza.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+ Within aigc enhancement, a new door opened. The blog was wroten based on virtual pizza shop data but demo a full process and customer satisfaction report and help to improve quality.
+
+This blog was first published at 
+[Enhancing Customer Satisfaction in Pizza Shops with RAG - A deep dive of Integrated pgai](https://github.com/timescale/pgai/blob/main/docs/pgai_rag_demo_why_customer_donot_like_pizza.md)
+
+## 7. Resources
+### 7.1 reference
+- [pgai](https://github.com/timescale/pgai)
+- [full demo sql used in this blog](https://github.com/timescale/pgaiexamples/rag_report_demo_why_some_customers_donot_like_pizza.sql)
+pizza dataset — it was copied from website, might be Gemini or OpenAI forum. But I can not find the original source link. After check with ChatGPT 4o, the text is common expression to pizza, not creative work.
```
[the remark this answers is not in the document]
Comment (Wang-Yong2018) [reply docs/pgai_rag_demo_why_customer_donot_like_pizza.md:277]: done

File: docs/pgai_rag_demo_why_customer_donot_like_pizza.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+Make sure the model connection and availability
+temporary set the api_key
+```sql
+set ai.openai_api_key = 'replace you api key here or use pgai default api_key environment';
+select pg_catalog.current_setting('ai.openai_api_key', true) as api_key;
+
+```
+
+### 3.2. Collect & Prepare Customer Voice 
+Listen to customer is key for long time success. Almost all business will listen to customer voice and records this. So, let create customer feedback data table and store the demo customer voices as record.
```
[the remark this answers is not in the document]
Comment (Wang-Yong2018) [reply docs/pgai_rag_demo_why_customer_donot_like_pizza.md:71]: done

File: docs/pgai_rag_demo_why_customer_donot_like_pizza.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+
+after few seconds, pgai generated a business report to you as below:
+```text
+# Business Report: Customer Feedback on Pizza
+
+## Summary of Customer Feedback
+
+Based on the collected feedback regarding our pizza offerings, several key points emerge that reveal customer dissatisfaction:
+
+1. **Quality Concerns**: A customer expressed extreme dissatisfaction, stating, \"The pizza was disgusting.\" This indicates potential issues with the ingredients or preparation. The mention of \"pepperoni made from rats\" suggests that there are serious concerns about the quality and safety of the ingredients used in our pizzas.
```
[the remark this answers is not in the document]
Comment (Wang-Yong2018) [reply docs/pgai_rag_demo_why_customer_donot_like_pizza.md:30]: thanks. I love those comments too.

File: docs/pgai_rag_demo_why_customer_donot_like_pizza.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+## 1. Executive Summary
+In this blog, we explore how integrating pgai with aigc and ChatGPT 4.0 can transform customer relationship management (CRM) in the pizza industry. By leveraging AI-driven insights, we demonstrate how to enhance customer satisfaction and operational efficiency in a virtual pizza shop scenario.
+
+## 2. Introduction
+The integration of Artificial Intelligence in General Computing (aigc) is revolutionizing business practices. This blog showcases the application of ChatGPT and pgai, an AI-enabled database, to improve customer interactions and obtain real-time insights.
+
+## 3 Case Study: Enhancing Customer Satisfaction in a Virtual Pizza Shop
+
+As a business analyst, you want to know why recently sell order drops from customer feedback database recenly and delivery a business report for management improvment. So, you give pgai a question by a SQL
```
[the remark this answers is not in the document]
Comment (Wang-Yong2018) [reply docs/pgai_rag_demo_why_customer_donot_like_pizza.md:17]: done. beside, I has used vscode spelling function to fix remain spelling issue too.

**Comment by Wang-Yong2018** [comment]
> I love this! Great work @Wang-Yong2018!
> 
> I think @contentana will love to port it to our official blog as well 🚀
> 
> A few ideas about naming:
> 
> 1. Clean up the suffixes and prefixes like `t_` and  `_v1` as they don't add real value to the case.
> 2. table names could be more meaningful, like `T_DEMO_TEXT_V1` can be `pizza_reviews` or simply `reviews` and the `subject` can be `Pizza`.
> 3. the `T_AI_REPORT` could also be like `ai_reports`
> 
> ```sql
> INSERT INTO reviews (product, message) VALUES
> 	 ('Pizza','The best pizza I''ve ever eaten. The sauce was so tangy!'),
> ```
> 
> Not asking you to change but just thinking loud that I love the potential of this example, and I'm already seeing we built something like an `ai_instructions` that could be a building block to not spread instructions around the entire database or keep it inside the functions. Something that could be:
> 
> ```sql
> insert into daily_reports (day, subject, content) as
> with top as (
>    select time_bucket('1 day', time),
>     string_agg(content) as content
>     FROM reviews where subject = 'Pizza Reviews' 
>     ORDER BY embedding <-> eqt.embedding as similarity
>     LIMIT 3
>   )
>  select build_report('daily operational issues report', top.content);
> ```

done, by new function build_pizza_report. hide all detail, just two input, 1. questioin, 2 n_criteria

**Review by jonatas — CHANGES_REQUESTED** [review]
Thanks for all the efforts @Wang-Yong2018!

Let's make it officially part of the Timescale blog! @contentana will move it forward from here using a google drive to store the final content.

I think the best for this PR is just hold the sql example file. Later we can add the link to the blog in the example file as a comment.

I'll reach out via DM to get your mail, so we can move the markdown content there 👍 


I'm going to reach out you to get your official mail so we can share a google drive folder with you and move it from there.

**Review by jonatas — APPROVED** [review]
I tested all of this locally and it works great!

I did some extra reports and used more data and it worked very well! I'm impressed 🚀 

```sql
select text_content from pizza_reviews_embeddings order by embedding <->  openai_embed('text-embedding-3-small', 'Satisfied customers');
text_content | I needed to call them twice and they're very friendly and helpful!
-------------+------------------------------------------------------------------------------------------
text_content | My favorite pizza flavor was not present but they created a special flavor for me!
-------------+------------------------------------------------------------------------------------------
text_content | The best pizza I've ever eaten. The sauce was so tangy!
-------------+------------------------------------------------------------------------------------------
text_content | The pizza delayed because it was late night, but it was delicious!
-------------+------------------------------------------------------------------------------------------
text_content | I loved the margherita traditional one. It was my favorite.
-------------+------------------------------------------------------------------------------------------
text_content | I ate 11 slices and threw up. The pizza was tasty in both directions.
-------------+------------------------------------------------------------------------------------------
text_content | I ordered a hot-dog and was given a pizza, but I ate it anyway.
-------------+------------------------------------------------------------------------------------------
text_content | I hate pineapple on pizza. It is a disgrace. Somehow, it worked well on this izza though.
-------------+------------------------------------------------------------------------------------------
text_content | The pizza was disgusting. I think the pepperoni was made from rats.

```

@Wang-Yong2018 please, double check to remove all this renaming references to allow us to move forward and merge it.


image

**Review by jgpruitt — CHANGES_REQUESTED** [review]

File: build.py
[the remark this answers is not in the document]
Comment (Wang-Yong2018) [reply build.py:1]: thanks for you feedback. 
How to do it? 
I added them in last submit, now delete them.  it looks like rework. 

  
the history:   
1 . these file was just copy from main branch. and for some how  it was added to root folder. 

Next:
1. Should I just  delete ai*.sql and build.py under rag_demo branch? 
I have delete the 4 files under rag_demo branch root folder.

2. should I using some git command to skip those files ? will those files delete bring new issue when merge? 
Best Regards,
WY
thanks.

**Pull request closed without merging**